File: IAIso-v5.0/core/iaiso-csharp/src/Iaiso.Middleware/Gemini.cs
using System.Collections.Generic;
using System.Text.Json.Nodes;
using Iaiso.Core;

namespace Iaiso.Middleware.Gemini;

/// <summary>IAIso wrapper for Google Gemini / Vertex AI generative models.</summary>
public interface IGeminiModel
{
    GeminiResponse GenerateContent(JsonObject request);
    string ModelName { get; }
}

public sealed class UsageMetadata
{
    public long PromptTokenCount { get; set; }
    public long CandidatesTokenCount { get; set; }
    public long TotalTokenCount { get; set; }
}

public sealed class Part
{
    public bool HasFunctionCall { get; }
    public Part(bool hasFunctionCall) { HasFunctionCall = hasFunctionCall; }
}

public sealed class Candidate
{
    public IReadOnlyList<Part> Parts { get; }
    public Candidate(IReadOnlyList<Part>? parts) { Parts = parts ?? new List<Part>(); }
}

public sealed class GeminiResponse
{
    public UsageMetadata UsageMetadata { get; }
    public IReadOnlyList<Candidate> Candidates { get; }
    public GeminiResponse(UsageMetadata? um, IReadOnlyList<Candidate>? candidates)
    {
        UsageMetadata = um ?? new UsageMetadata();
        Candidates = candidates ?? new List<Candidate>();
    }
}

public sealed class GeminiOptions
{
    public bool RaiseOnEscalation { get; set; }
    public static GeminiOptions Defaults() => new();
}

public sealed class BoundedGeminiModel
{
    private readonly IGeminiModel _raw;
    private readonly BoundedExecution _execution;
    private readonly GeminiOptions _opts;

    public BoundedGeminiModel(IGeminiModel raw, BoundedExecution execution, GeminiOptions? opts = null)
    {
        _raw = raw;
        _execution = execution;
        _opts = opts ?? GeminiOptions.Defaults();
    }

    public GeminiResponse GenerateContent(JsonObject request)
    {
        var pre = _execution.Check();
        if (pre == StepOutcome.Locked) throw new MiddlewareException.Locked();
        if (pre == StepOutcome.Escalated && _opts.RaiseOnEscalation)
            throw new MiddlewareException.EscalationRaised();

        GeminiResponse resp;
        try { resp = _raw.GenerateContent(request); }
        catch (System.Exception e) { throw new MiddlewareException.Provider(e.Message, e); }

        long tokens = resp.UsageMetadata.TotalTokenCount;
        if (tokens == 0)
            tokens = resp.UsageMetadata.PromptTokenCount + resp.UsageMetadata.CandidatesTokenCount;
        long toolCalls = 0;
        foreach (var c in resp.Candidates)
            foreach (var p in c.Parts) if (p.HasFunctionCall) toolCalls++;
        string model = string.IsNullOrEmpty(_raw.ModelName) ? "unknown" : _raw.ModelName;
        _execution.RecordStep(new StepInput
        {
            Tokens = tokens,
            ToolCalls = toolCalls,
            Tag = $"gemini.generateContent:{model}",
        });
        return resp;
    }
}

File: IAIso-v5.0/core/iaiso-swift/Sources/IAIsoMiddleware/Gemini.swift
import Foundation
import IAIsoCore

/// Google Gemini / Vertex AI generative models middleware.
public enum Gemini {

    public protocol Model {
        func generateContent(_ request: [String: Any]) throws -> Response
        var modelName: String { get }
    }

    public struct UsageMetadata: Sendable, Equatable {
        public let promptTokenCount: Int64
        public let candidatesTokenCount: Int64
        public let totalTokenCount: Int64
        public init(
            promptTokenCount: Int64 = 0,
            candidatesTokenCount: Int64 = 0,
            totalTokenCount: Int64 = 0
        ) {
            self.promptTokenCount = promptTokenCount
            self.candidatesTokenCount = candidatesTokenCount
            self.totalTokenCount = totalTokenCount
        }
    }

    public struct Part: Sendable, Equatable {
        public let hasFunctionCall: Bool
        public init(hasFunctionCall: Bool = false) {
            self.hasFunctionCall = hasFunctionCall
        }
    }

    public struct Candidate: Sendable {
        public let parts: [Part]
        public init(parts: [Part] = []) { self.parts = parts }
    }

    public struct Response: Sendable {
        public let usageMetadata: UsageMetadata
        public let candidates: [Candidate]
        public init(usageMetadata: UsageMetadata, candidates: [Candidate]) {
            self.usageMetadata = usageMetadata
            self.candidates = candidates
        }
    }

    public struct Options: Sendable {
        public var raiseOnEscalation: Bool
        public init(raiseOnEscalation: Bool = false) {
            self.raiseOnEscalation = raiseOnEscalation
        }
        public static let defaults = Options()
    }

    public final class BoundedModel {
        private let raw: Model
        private let execution: BoundedExecution
        private let opts: Options

        public init(raw: Model, execution: BoundedExecution, options: Options = .defaults) {
            self.raw = raw
            self.execution = execution
            self.opts = options
        }

        public func generateContent(_ request: [String: Any]) throws -> Response {
            let pre = execution.check()
            if pre == .locked { throw MiddlewareError.locked }
            if pre == .escalated, opts.raiseOnEscalation {
                throw MiddlewareError.escalationRaised
            }
            let resp: Response
            do {
                resp = try raw.generateContent(request)
            } catch {
                throw MiddlewareError.provider(
                    error.localizedDescription, underlying: error)
            }
            var tokens = resp.usageMetadata.totalTokenCount
            if tokens == 0 {
                tokens = resp.usageMetadata.promptTokenCount
                    + resp.usageMetadata.candidatesTokenCount
            }
            var toolCalls: Int64 = 0
            for c in resp.candidates {
                for p in c.parts {
                    if p.hasFunctionCall { toolCalls += 1 }
                }
            }
            let model = raw.modelName.isEmpty ? "unknown" : raw.modelName
            execution.recordStep(StepInput(
                tokens: tokens,
                toolCalls: toolCalls,
                tag: "gemini.generateContent:\(model)"))
            return resp
        }
    }
}
